sssd_test_framework/utils/adcli.py

```diff
@@ -0,0 +1,80 @@
+"""Perform actions on Active Directory."""
+
+from __future__ import annotations
+
+from pytest_mh import MultihostHost, MultihostUtility
+from pytest_mh.cli import CLIBuilder, CLIBuilderArgs
+from pytest_mh.conn import ProcessResult
+from pytest_mh.utils.fs import LinuxFileSystem
+
+__all__ = [
+    "ADCLIUtils",
+]
+
+
+class ADCLI(MultihostUtility[MultihostHost]):
+    """
+    Call commands from adcli
+    """
+
+    def __init__(self, host: MultihostHost, fs: LinuxFileSystem) -> None:
+        super().__init__(host)
+
+        self.cli: CLIBuilder = self.host.cli
+        """Command line builder."""
+
+        self.fs: LinuxFileSystem = fs
+        """Filesystem utils."""
+
+    def info(
+        self,
+        domain: str,
+    ) -> str:
+        """
+        Call ``adcli info `` with given arguments.
+
+        :param domain: Displays discovered information about an Active Directory domain, defaults to None
+        :type domain: str | None,
+        :param domain_controller: Domain controller to connect, defaults to None
+        :type domain_controller: str | None, optional
+        """
+        args: CLIBuilderArgs = {
+            "domain": (self.cli.option.POSITIONAL, domain),
+            "domain_controller": (self.cli.option.POSITIONAL, domain_controller),
+        }
+
+        return self.host.conn.exec(["adcli", "info"] + self.cli.args(args))
+
+    def testjoin(
+        self,
+        domain: str| None = None,
+    ) -> ProcessResult:
+        """
+        call ``adcli testjoin `` with given arguments.
+
+        :param domain: Target Active Directory domain, defaults to None
+        :type domain: str | None, optional
+        :param domain_controller: Domain controller to connect
+        :type domain_controller: str | None, optional
+        """
+        args: CLIBuilderArgs = {
+            "domain": (self.cli.option.VALUE, domain),
+            "domain_controller": (self.cli.option.VALUE, domain_controller),
+            "host_keytab": (self.cli.option.VALUE, host_keytab),
+        }
+        return self.host.conn.exec(["adcli", "testjoin"] + self.cli.args(args))
+
+    def join(
+        self,
+        domain: str,
+    ) -> str:
```

Comment:
Could we need to run join with additional options/arguments like computer name, host keytab, user principal or others?   If we do, I think we might be able to use something like the arg list used in utils/tools.py in SSHKeyUtils.generate here:

https://github.com/SSSD/sssd-test-framework/blob/master/sssd_test_framework/utils/tools.py#L770

Reply:
Does the change work?   I'd think you'd need to keep the domain inside the CLIBuilderArgs to do something like:

```
        if args is None:
            args = []

        bargs: CLIBuilderArgs = {
            "domain": (self.cli.option.POSITIONAL, domain),
        }
        self.host.conn.exec(["adcli", "join"] + self.cli.args(bargs) + [*args])
```

I haven't tested the above but, I'm just trying to mirror what was done in the SSHKeyUtils.generate().  Also, I'm not sure of the adcli command order.   Maybe args should be right after the join?

Reply:
I think the change with Pavel's suggested update resolves my questions for this one.   marking resolved.